# Perturbations in Intestinal Antibiotic-resistant Bacterial Populations in Dairy Cattle Following Intramammary Treatment Using a Third-generation Cephalosporin 

## Abstract

Background. The impact of antimicrobials used to treat and prevent diseases in livestock requires careful monitoring, particularly for drugs considered to be critically important for human medicine. Such is the case of third-generation cephalosporins, which are commonly used in cattle farms. Use of these drugs may select for antibiotic resistant (AR) bacteria including Extended-spectrum β-lactamase (ESBL)-producing Enterobacteriaceae, the most economically impactful antimicrobial-resistant threat. Hence, we sought to evaluate how intramammary (IMM) treatment with ceftiofur impacts the cattle-gut microbiome and the abundance of resistant bacteria.

Methods. Forty healthy dairy cows were enrolled at the end of lactation; 20 were treated with IMM ceftiofur (Spectramast®DC) and a non-antibiotic internal teat sealant (Orbeseal®) and 20 received only Orbeseal® as a control. Fecal samples were collected from each cow before (day -1) and after treatment (weeks 1, 2, 3, 5, 7 & 9) for quantification of bacteria with resistance to β-lactams as well as metagenomic next generation sequencing (mNGS). Gram-negative and -positive bacteria were quantified on MacConkey-lactose and Columbia CNA agar, respectively, while the initial mNGS and bioinformatic analyses were done using CosmosID pipelines.

<p align="center">
  <img src="https://github.com/karla-vasco/metagenome_cows_IMM-ceftiofur/blob/main/Study_methods.png?raw=true" />
</p>

Results. One week after dry-off, cows treated with IMM ceftiofur had 3-fold fewer Gram-negative bacteria (P = 0.006) and 2-fold fewer Gram-positive bacteria than controls (P = 0.066), while there were no differences in the quantity of AR bacteria based on treatment. Although there was no difference in the prevalence of bacteria resistant to β-lactams between treatment groups, 90% and 24% of samples analyzed (n = 278) had Gram-negatives resistant to ampicillin and ceftiofur, respectively. On average, the number of Gram-negative ceftiofur-resistant bacteria increased for up to 2 weeks in the cows treated with antibiotic, though it was not significant due to a high variation among animals where 25% of them shed more ceftiofur-resistant bacteria. While IMM ceftiofur treatment did not affect the richness or composition of the microbiome or resistome, the number of bacteria from several taxa decreased one week after treatment compared to controls. Notably, fecal samples from cows treated with ceftiofur persistently had a greater abundance of ESBL genes (bla-CFXA and bla-CMY) throughout the 9-week period (LDA > 3.3; P < 0.03) that were associated with Enterobacteriaceae and commensal bacteria. Network analysis also showed a persistent increase of correlations between β-lactam resistant genes with phages, MGEs, and bacterial genera during weeks 1 and 9 post IMM ceftiofur treatment.

Conclusion. This study highlights the long-term effects of 2 g IMM ceftiofur given at dry off on the gut microbiome and resistome composition in dairy cattle. Despite a single IMM administration of a third-generation cephalosporin, treated cattle increased the abundance of genes encoding ESBLs in feces that persisted for 9 weeks. Commensal bacteria were mainly associated with carriage of ESBLs and in lower proportion Enterobacteriaceae. Shedding of ESBL-Enterobacteriaceae was variable among animals, where 75% did not increase the number of ceftiofur-resistant bacteria after IMM antibiotic application. Additional research is therefore needed to identify associations between fecal shedding and the potential for dissemination and persistence of antibiotic resistance determinants in dairy farms.

## Metagenome analysis

<p align="center">
  <img src="https://github.com/karla-vasco/metagenome_cows_IMM-ceftiofur/blob/main/Metagenome_analysis.png?raw=true" />
</p>
