---
title: "Virome profiling"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```

# Activating Libraries
```{r error=FALSE}
library(readxl)
library(ggpubr)
library(ggsci)
library(dplyr)
library(rstatix)

# Function use to filter data that do not contain an element or list of elements with dplyr
`%notin%` <- Negate(`%in%`)
```

# Data wrangling

Importing virome abundance table
```{r}
library(readr)

setwd("/Users/karlavasco/Library/CloudStorage/OneDrive-MichiganStateUniversity/Manning_lab/Mastitis_project/metagenome_KV/tables/virome/")
virome <- read_tsv("virushost_kma_seqname_noheader.tsv",col_names = c("Template",	"Score",	"Expected",	"Template_length",	"Template_Identity",	"Template_Coverage",	"Query_Identity",	"Query_Coverage",	"Depth",	"q_value",	"p_value","sequence_id"))
```

Importing sample data
```{r}
setwd("/Users/karlavasco/Library/CloudStorage/OneDrive-MichiganStateUniversity/Manning_lab/Mastitis_project/metagenome_KV/tables/metadata/")
data <- read_excel("IMM_ceftiofur_metadata.xlsx", sheet = "metadata")
temperature <- read_excel("IMM_ceftiofur_metadata.xlsx", sheet = "temperature")
diet <- read_excel("IMM_ceftiofur_metadata.xlsx", sheet = "diet")
metagenome_stats <- read_excel("IMM_ceftiofur_metadata.xlsx", sheet = "metagenome_stats")
```

Merging tables for metadata
```{r}
merged <- merge(data, diet) %>% 
  merge(temperature) %>%
  merge(metagenome_stats) %>% 
  merge(virome) %>% 
  mutate(Treatment = factor(Treatment, levels = c("Control", "Antibiotic")),
          Time_tx = factor(Time_tx, levels = c("Day -1", "Week 1", "Week 5", "Week 9")))
```

New column with normalized abundance based on Genome equivalents
```{r}
merged <- merged %>% mutate(normalized_abundance = Depth/genome_equivalents*100)
```

# Linear-mixed effect models
```{r}
library(nlme)
lm.test <- lme(Depth ~Time_tx+Time_tx*temperature_Celsius+Time_tx*Treatment, random=~1|study_ID, data = merged)
summary(lm.test)
anova(lm.test)
```


# Virome abundance

Virome Abundance table
```{r}
abundance_feature <- merged %>% 
  select(normalized_abundance, Treatment, Time_tx, sample_ID) %>% 
  group_by(Treatment, Time_tx, sample_ID) %>% 
  dplyr::summarise(Abundance = sum(normalized_abundance)) %>% as.data.frame() %>% mutate(Type="Viruses")

abundance_feature$Treatment <- factor(abundance_feature$Treatment, levels= c("Control","Antibiotic"))

```

Abundance plot
```{r}
#Calculating p-values between treatments by time point
stat.test <- abundance_feature %>% 
    dplyr::filter(sample_ID %notin% c("MA028.7")) %>% 
  group_by(Time_tx) %>%
  wilcox_test(Abundance ~ Treatment, alternative = "less", paired = T) %>%
  adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj") %>%
  add_xy_position(x = "Time_tx", dodge = 0.8)
stat.test

#Boxplot of Observed diversity between treatment groups over time
Abundance_virus = abundance_feature %>%  
  ggboxplot(x = "Time_tx", y = "Abundance", color = "Treatment", palette = "jama", fill = "Treatment", add = c("jitter"), notch = F, outlier.shape = NA,facet.by = "Type", ylab = F, xlab = F) +
  scale_fill_jama(alpha = 0.5) +
  ylim(1,18)+
  stat_pvalue_manual(stat.test,  label = "p", tip.length = 0) +  theme(axis.title.x = element_blank())

Abundance_virus
```

# Virome diversity

## Phyloseq table formatting

Feature table used for alpha diversity: Matrix with sequence_id as columns and features as rows
```{r}
library(tibble)
feature_alpha <- merged %>% select(sequence_id,Template,Depth) %>% 
  spread(sequence_id, Depth) %>% 
  remove_rownames %>% column_to_rownames(var="Template") %>% 
  as.matrix(rownames=TRUE)
feature_alpha[is.na(feature_alpha)] <- 0
```

Taxonomy table
```{r}
taxonomy <- virome %>% select(Template) %>% distinct() %>%
  mutate(Feature = Template) %>% 
  remove_rownames %>% column_to_rownames(var="Template") %>% 
  as.matrix()
```

Metadata to matrix
```{r}
metadata_matrix <- merge(data, metagenome_stats) %>% remove_rownames %>% column_to_rownames(var="sequence_id")
```

Phyloseq object
```{r}
library("phyloseq")
feature_alpha <- feature_alpha*10^4
mode(feature_alpha) <- "integer"
feature_alpha[is.na(feature_alpha)] <- 0


#import as phyloseq objects
OTU = otu_table(feature_alpha,taxa_are_rows=TRUE)
TAX = tax_table(taxonomy)
META = sample_data(metadata_matrix)
#(tree was already imported as a phyloseq object)
physeq=phyloseq(OTU,TAX,META)
```

## Alpha diversity
```{r fig.width=10, fig.height=40}
alpha_diversity <- estimate_richness(physeq, measures = c("Shannon", "Observed", "Chao1"))
df_alpha <- data.frame(alpha_diversity, sample_data(physeq))
alpha_table <- reshape2::melt(df_alpha, measure.var=c("Shannon","Observed","Chao1"),id.vars=c("Time_tx","Treatment","sample_ID","study_ID","Block","Cohort")) %>% mutate(Type="Viruses") %>% 
  dplyr::rename(Index = variable)
alpha_table$value = as.numeric(alpha_table$value)
alpha_table$Treatment <- factor(alpha_table$Treatment, levels= c("Control","Antibiotic"))
```

Shannon diversity
```{r}
#Calculating p-values between treatments by time point
stat.test <- alpha_table %>% 
  filter(Index == "Shannon",sample_ID %notin% c("MA028.7")) %>% 
  group_by(Time_tx) %>%
  wilcox_test(value ~ Treatment, alternative = "less", paired = T) %>%
  adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj") %>%
  add_xy_position(x = "Time_tx", dodge = 0.8)
stat.test

#Boxplot of Shannon diversity between treatment groups over time
shannon_virus = alpha_table %>% 
  filter(Index == "Shannon") %>% 
  ggboxplot(x = "Time_tx", y = "value", color = "Treatment", palette = "jama", fill = "Treatment", add = c("jitter"), notch = F, outlier.shape = NA,facet.by = "Type") +
  labs(x = "Time to treatment", y = "Shannon index", colour = "Treatment", fill = "Treatment") + 
  scale_fill_jama(alpha = 0.5) +
  ylim(0,3.2)+
  stat_pvalue_manual(stat.test,  label = "p", tip.length = 0) +  theme(axis.title.x = element_blank())

shannon_virus
```

Observed diversity
```{r}
#Calculating p-values between treatments by time point
stat.test <- alpha_table %>% 
  filter(Index == "Observed",sample_ID %notin% c("MA028.7")) %>% 
  group_by(Time_tx) %>%
  wilcox_test(value ~ Treatment, alternative = "less", paired = T) %>%
  adjust_pvalue(method = "bonferroni") %>%
  add_significance("p.adj") %>%
  add_xy_position(x = "Time_tx", dodge = 0.8)
stat.test

#Boxplot of Observed diversity between treatment groups over time
Observed_virus = alpha_table %>% 
  filter(Index == "Observed") %>% 
  ggboxplot(x = "Time_tx", y = "value", color = "Treatment", palette = "jama", fill = "Treatment", add = c("jitter"), notch = F, outlier.shape = NA,facet.by = "Type") +
  labs(x = "Time to treatment", y = "Observed", colour = "Treatment", fill = "Treatment") + 
  scale_fill_jama(alpha = 0.5) +
  ylim(0,55)+
  stat_pvalue_manual(stat.test,  label = "p", tip.length = 0) +  theme(axis.title.x = element_blank())

Observed_virus
```

## Beta diversity 
Feature table used for beta diversity: Matrix with sequence_id as columns and features as rows
```{r}
library(tibble)
feature_table <- merged %>% select(sequence_id,Template,normalized_abundance) %>% 
  spread(sequence_id, normalized_abundance) %>% 
  remove_rownames %>% column_to_rownames(var="Template") %>% 
  as.matrix(rownames=TRUE)
feature_table[is.na(feature_table)] <- 0
```

Phyloseq object for beta diversity
```{r}
library("phyloseq")
#import as phyloseq objects
OTU = otu_table(feature_table,taxa_are_rows=TRUE)
#(tree was already imported as a phyloseq object)
physeq=phyloseq(OTU,TAX,META)
```

PERMANOVA
```{r}
library(microbial)
beta <-betatest(physeq,group="Time_tx")
beta
```

Bray-Curtis plot
```{r fig.width=8}
bray_virome <- plotbeta(
  physeq,
  group="Time_tx",
  shape = "Treatment",
  distance = "bray",
  method = "PCoA",
  color = NULL,
  size = 3,
  ellipse = F) +
    labs(color = "Time", shape = "Treatment", fill="Time") + 
  annotate("text", x = 0, y = 0.2, label = expression(paste("PERMANOVA, ",F ,"= 9.2, ",paste(italic('p')),"=0.001")), colour = "black", size = 4) + ggtitle("Viruses") +
  stat_ellipse(geom = "polygon",
               aes(fill = Time_tx, group=Time_tx), 
               alpha = 0.1)+
  scale_fill_aaas() +
  scale_color_aaas()+
  theme_bw()

bray_virome
```

Virome diversity plot
```{r fig.width = 10, fig.height= 3}
alpha_virus <- ggarrange(shannon_virus, Observed_virus, common.legend = T, nrow = 1)
div_virus <- ggarrange(alpha_virus, bray_virome, nrow = 1, labels = c("C"))
div_virus
```

```{r results='hold'}
#accessory_ab_plots <- ggarrange(Abundance_plas, Abundance_vf, Abundance_virus, nrow = 1, common.legend = T)
#accessory_ab_plots
```

```{r results='hold'}
#setwd("/Users/karlavasco/Library/CloudStorage/OneDrive-MichiganStateUniversity/Manning_lab/Mastitis_project/metagenome_KV/figures")
#ggsave(plot=accessory_ab_plots, "Fig7_accessory_abundance.png", width = 14, height = 4, bg = "white")
```

```{r}
#accessory_plots <- ggarrange(div_plas, div_vf, div_virus, nrow = 3)
#accessory_plots
```

```{r}
#setwd("/Users/karlavasco/Library/CloudStorage/OneDrive-MichiganStateUniversity/Manning_lab/Mastitis_project/metagenome_KV/figures")
#ggsave(plot=accessory_plots, "FigS5_accessory_diversity.png", width = 14, height = 12, bg = "white")
```
