---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---

#ANCOM-BC
https://github.com/FrederickHuangLin/ANCOM-BC-Code-Archive

##Installing ANCOMBC if necessary
```{r}
if (!requireNamespace("BiocManager", quietly=TRUE))
    install.packages("BiocManager")
BiocManager::install("ANCOMBC")
```

##Activating libraries
```{r}
library(microbiome)
library(nloptr)
library(ANCOMBC)
```

##Importing tables
```{r echo = FALSE}
library(dplyr)

setwd("/Users/karlavasco/OneDrive\ -\ Michigan\ State\ University/Manning_lab/Mastitis_project/GitHub/data/")

#read in otu table
otu_table = read.csv("feature_table_genes.csv",sep=",", check.names = F, row.names = 1)
otu_table=as.matrix(otu_table)


#read in taxonomy
taxonomy = read.csv("taxonomy_genes.csv",sep=",", row.names = 1)
taxonomy=as.matrix(taxonomy)

#read in metadata
metadata=read.table("metadata.csv", sep = ",", header = TRUE, row.names=1, na.strings = "") 
metadata
```

##Generating a phyloseq object
```{r echo=FALSE}
library("phyloseq")

#import as phyloseq objects
OTU = otu_table(otu_table,taxa_are_rows=TRUE)
TAX = tax_table(taxonomy)
META = sample_data(metadata)

# Sanity checks for consistent OTU names
taxa_names(TAX)
taxa_names(OTU)

# Same sample names
sample_names(OTU)
sample_names(META)

#Final phyloseq output is named as physeq
physeq=phyloseq(OTU,TAX,META)
```

##Filtering samples per time point
```{r}
physeq_day1 <- subset_samples(physeq, Time_tx%in%"Day -1")
physeq_week1 <- subset_samples(physeq, Time_tx%in%"Week 1")
physeq_week5 <- subset_samples(physeq, Time_tx%in%"Week 5")
physeq_week9 <- subset_samples(physeq, Time_tx%in%"Week 9")
```

#ANCOMBC DAY -1

##Type: ARG, MGEs or Virulence genes -> between treatment groups
```{r}
# Aggregate the abundance to the "Gene Type" level
physeq = aggregate_taxa(physeq_day1, "Type")

#ANCOMBC analysis comparison between treatment groups
out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)

#ANCOMBC results as a list
res = out$res

#ANCOMBC results as a table
ancom_results = res %>% 
  as_tibble()

#Filtering only significant results
ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)

#Adding feature names to the table
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Type")

#Exporting significant results as a csv file
write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_type_day1.csv",sep = ",", row.names = F)
```

##Class between treatment groups
```{r}
# Aggregate the abundance to the "Gene Class" level
physeq = aggregate_taxa(physeq_day1, "Class")

#ANCOMBC analysis comparison between treatment groups
out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)

#ANCOMBC results as a list
res = out$res

#ANCOMBC results as a table
ancom_results = res %>% 
  as_tibble()

#Filtering only significant results
ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)

#Adding feature names to the table
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Class")

#Exporting significant results as a csv file
write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_class_day1.csv",sep = ",", row.names = F)
```


##Mechanism between treatment groups
```{r}
# Aggregate the abundance to the "Gene Mechanism" level
physeq = aggregate_taxa(physeq_day1, "Mechanism")

#ANCOMBC analysis comparison between treatment groups
out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)

#ANCOMBC results as a list
res = out$res

#ANCOMBC results as a table
ancom_results = res %>% 
  as_tibble()

#Filtering only significant results
ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)

#Adding feature names to the table
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Mechanism")

#Exporting significant results as a csv file
write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_mechanism_day1.csv",sep = ",", row.names = F) 
```

##Gene
```{r}
# Aggregate the abundance to the "Gene" level
physeq = aggregate_taxa(physeq_day1, "Gene")

#ANCOMBC analysis comparison between treatment groups
out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)

#ANCOMBC results as a list
res = out$res

#ANCOMBC results as a table
ancom_results = res %>% 
  as_tibble()

#Filtering only significant results
ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)

#Adding feature names to the table
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Gene")

#Exporting significant results as a csv file
write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_gene_day1.csv",sep = ",", row.names = F) 
```

##Name
```{r}
# Aggregate the abundance to the "Name" level, it has data for virulence genes where name corresponds to an associated bacterial species

physeq = aggregate_taxa(physeq_day1, "Name")

#ANCOMBC analysis comparison between treatment groups
out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)

#ANCOMBC results as a list
res = out$res

#ANCOMBC results as a table
ancom_results = res %>% 
  as_tibble()

#Filtering only significant results
ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)

#Adding feature names to the table
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Name")

#Exporting significant results as a csv file
write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_name_day1.csv",sep = ",", row.names = F)
```

#WEEK 1
##Type
```{r}
physeq = aggregate_taxa(physeq_week1, "Type")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Type")

#write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_type_week1.csv",sep = ",", row.names = F) #Writing `sigtab` to
```

##Class
```{r}
# Aggregate to Class level
physeq = aggregate_taxa(physeq_week1, "Class")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Class")

#write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_class_week1.csv",sep = ",", row.names = F) #Writing `sigtab` to
```


##Mechanism
```{r}
# Aggregate to mechanism level
physeq = aggregate_taxa(physeq_week1, "Mechanism")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Mechanism")

write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_mechanism_week1.csv",sep = ",", row.names = F)
```

##Gene
```{r}
# Aggregate to Gene level
physeq = aggregate_taxa(physeq_week1, "Gene")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Gene")

write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_gene_week1.csv",sep = ",", row.names = F) 
```

#WEEK 5
##Type
```{r}
# Aggregate to type level
physeq = aggregate_taxa(physeq_week5, "Type")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Type")

#write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_type_week5.csv",sep = ",", row.names = F) #Writing `sigtab` to
```

##Class
```{r}
# Aggregate to Class level
physeq = aggregate_taxa(physeq_week5, "Class")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Class")

#write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_class_week5.csv",sep = ",", row.names = F) #Writing `sigtab` to
```


##Mechanism
```{r}
# Aggregate to mechanism level
physeq = aggregate_taxa(physeq_week5, "Mechanism")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Mechanism")

write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_mechanism_week5.csv",sep = ",", row.names = F)
```

##Gene
```{r}
# Aggregate to Genus level
physeq = aggregate_taxa(physeq_week5, "Gene")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Gene")

write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_gene_week5.csv",sep = ",", row.names = F) 
```

#WEEK 9
##Type
```{r}
# Aggregate to Type level
physeq = aggregate_taxa(physeq_week9, "Type")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Type")

#write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_type_week9.csv",sep = ",", row.names = F) #Writing `sigtab` to
```

##Class
```{r}
# Aggregate to Class level
physeq = aggregate_taxa(physeq_week9, "Class")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Class")

#write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_class_week9.csv",sep = ",", row.names = F) #Writing `sigtab` to
```


##Mechanism
```{r}
# Aggregate to mechanism level
physeq = aggregate_taxa(physeq_week9, "Mechanism")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Mechanism")

#write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_mechanism_week9.csv",sep = ",", row.names = F) #Writing `sigtab` to
```

##Gene
```{r}
# Aggregate to Gene level
physeq = aggregate_taxa(physeq_week9, "Gene")

out = ancombc(phyloseq = physeq, formula = "Treatment",
              p_adj_method = "holm", zero_cut = 0.90, lib_cut = 0,
              group = "Treatment", struc_zero = F, neg_lb = FALSE,
              tol = 1e-05, max_iter = 100, conserve = TRUE,
              alpha = 0.05, global = TRUE)
res = out$res

ancom_results = res %>% 
  as_tibble()

ancom_signif_list <- ancom_results %>% 
  filter(p_val <= 0.05)
ancom_signif_names <- data.frame(sapply(ancom_signif_list,c)) 
row.names(ancom_signif_names) <- rownames(ancom_signif_list$beta)
ancom_signif_names <- tibble::rownames_to_column(ancom_signif_names, "Gene")

write.table(ancom_signif_names, file="~/OneDrive - Michigan State University/Manning_lab/Mastitis_project/tables/diff_abundance/ANCOMBC/genes/ANCOM_gene_week9.csv",sep = ",", row.names = F) 
```

